### PR TITLE
feat: finish populating existing precompiles and predeploys

### LINKED ISSUE #3: feat: finish populating existing precompiles and predeploys (bounty: 75 OP 🔴)
*This issue is eligible for a 75 OP bounty. Read [CONTRIBUTING.md](https://github.com/mds1/evm-diff/blob/main/CONTRIBUTING.md) to learn how to qualify.*

This is basically a version of https://www.evm.codes/precompiled for each chain, with the addition of predeploys. This has been started but is not left complete. The remaining tasks are:

1. Search for `TODO` in any `precompiles.ts` files to reveal the uncompleted sections, and complete them.
2. All existing `references` fields should be checked for accuracy
3. Additional references should be added for more information when possible.

### PR DESCRIPTION
**Pull request by cairoeth**
Completed TODOs around precompiles for Arbitrum and Mainnet. Also added missing references for Mainnet precompiles and checked Optimism's. Resolves #3 

Optimism payout address: 0x741cB6A6a8dC16363666462769D8dEc996311466

### CODE REVIEW

File: src/chains/mainnet/vm/precompiles.ts [hunk 1]
```diff
@@ -91,60 +92,286 @@ export const precompiles: Precompile[] = [
...
+      },
+      {
+        byteStart: 96,
+        byteLength: 96 + 32,
+        name: 'B',
+        description: 'Base as unsigned integer',
+      },
+      {
+        byteStart: 96 + 32,
+        byteLength: 96 + 32 + 32,
```
Comment (cairoeth) [inline src/chains/mainnet/vm/precompiles.ts:156]: The byte length of B, E, M and value (output) are dependant on each other, so not sure how to make it variable length. maybe with -1?

**Comment by mds1** [comment]
Thanks for the quick turnaround here, planning to review this tomorrow

File: src/chains/mainnet/vm/precompiles.ts (hunk 1 above)
Comment (mds1) [reply src/chains/mainnet/vm/precompiles.ts:156]: Hmm, I'd say in the `PrecompileParam`, let's change the types of both `byteStart` and `byteLength` from `number` to `number | string`, and we can write this out similar to https://www.evm.codes/precompiled#0x05?fork=shanghai, e.g. 

```typescript
{
  byteStart: '96 + Bsize',
  byteLength: 'Esize'
}
```

Then the `PrecompileDiff` component can check the type and format things accordingly, and we can show more details this way than if we just used `-1`. (For other precompiles that have a single, variable length input, using -1 seems ok, but here it's a bit more complex)

File: src/chains/mainnet/vm/precompiles.ts [hunk 2]
```diff
@@ -68,7 +67,9 @@ export const precompiles: Precompile[] = [
         description: 'The SHA-256 hash of the input data',
       },
     ],
-    references: [],
+    references: [
+      'https://github.com/ethereum/execution-specs/blob/master/src/ethereum/shanghai/vm/precompiled_contracts/sha256.py',
```
Comment (mds1) [inline src/chains/mainnet/vm/precompiles.ts:71]: For each mainnet precompile could you include all 3 `ethereum/execution-specs` references, e.g. the ecReover precompile links to the files in the execution-spects defining it's (1) implementation, (2) min gas cost, and (3) address, which are in different files

File: src/chains/mainnet/vm/precompiles.ts [hunk 3]
```diff
@@ -91,60 +92,280 @@ export const precompiles: Precompile[] = [
...
+      {
+        byteStart: 128,
+        byteLength: 32,
+        name: 'y3',
+        description: "Y coordinate of the third point on the elliptic curve 'alt_bn128'",
+      },
+      {
+        byteStart: 160,
+        byteLength: 32,
+        name: 'y2',
```
Comment (mds1) [inline src/chains/mainnet/vm/precompiles.ts:307]: I think the order got swapped here (it goes x1,y1,x3,x2,y3,y2). Let's keep the first two as `x1`,`y1` and `x2,y2`, then make the last one `xn`,`yn` to indicate it's arbitrary length number of inputs

**Review by mds1 — COMMENTED** [review]
This looks great, thanks! Just a few small comments

**Review by mds1 — APPROVED** [review]
Thank you! https://optimistic.etherscan.io/tx/0x26879fcb8f7a8f1d106377e1f6a6a31bc13f8097981b0288d5defdebbdcc7daa

**Pull request merged**